### DISCUSSION TITLE
[5/13] Knowledge Distillation

### DISCUSSION THREAD

**Discussion by tjdtnsu** [Pseudo-Lab/EfficientML#21] [General]
.

**Comment by namgyu-youn** [comment]
## Challenge : Compression methods (i.e., Pruning, Quantization)
- Compression based approach needs parent (base) model. And it is too high to run those model
- This limitation have motivated to find fundamental tiny architecture, and it became NAS and KD

## Knowledge Distillation Basics
- Main concept : Teacher (heavy) model helps building student (light) model by using their logits
- Logits : The kind of output. It can be GT, channel's output, etc.
- Distillation Loss : Evaluation metrics for checking if student successfully learned from Teacher model.

## Matching Intermediate Feature & Weights
- Recap) Activation has high-memory cost issue since it is non-linear, and memory-bound. (c.f., Mixed Precision)
- KD forces imitating Teacher's middle layer (feature map & activation) (Middle layers are applied to many theories, like Layer-CAM)

## Matching Intermediate Attention Maps
- Attention value directly related to impact in feature map
- Although high-attention value zone is not common since model's performance differs, **low-attention value zone is commonly same**
- Note: The goal is minimizing gap between `P(t)` and `P(s)`
- Relationship between **T and S could be collapsed when dataset is changed** -> Feature map's distribution is core feature

## Born Again Neural networks (BANs) ([Blog](https://until.blog/@namgyu-youn/-icml2018--born-again-neural-networks))
- Q. Should we use Teacher while building Student's architecture? -> BANs introduces self-validation based Student for this issue
- I.A : Stochastic based self-validation and self-ensemble were verified (c.f., K-fold cross validation)
- Ground Truth (GT) : Importance weight preserving only answer (gt) and simply rescalable
- Conclusion : BANs firstly overcome Teacher using self-distillation. (It was speciallized in image classification task)

## KD for Object Detection & Segmentation & NLP
- Key point is, we should build **"appropriate distillation loss"** for applying the task.
- Student should aim to follow Teacher, **not overtake**. For restricting it, we can use upper bound

## Network Augmentation
- Like data augmentation in small data sample task, we can start with smallest model without teacher.
- NetAug : Gradually increase Student architecture's size
- In final project, sLLM uses NetAug (interesting...)

**Comment by Saebitna** [comment]
**What is Knowledge distillation (KD)**
- Align the class probability distributions from teacher and student networks (using distillation loss)

**What to match**
- output logit
- intermediate weights
- intermediate features
- gradients of feature maps (intermediate attention maps)
- sparsity patterns (after the ReLU activation)
- relational information: (1) relations between different layers (inner project) (2) relations between different samples (pairwise distances of feature vectors)

**Self and Online Distillation**
- Self distillation > Born-Again NN: *외부* teacher 없이 같은 네트워크 혹은 동일 구조의 반복 학습만으로 지식을 순환‧재전파해 성능을 끌어올리는 방법
- Online distillation > Deep Mutual Learning: peers teach each other during training
- Combined > Be Your Own Teacher:

**Application Tasks**
- Object Detection > Feature Imitation, localization distillation
- Semantic Segmentation (pixel-wise prediction) > KD + discriminator network
- KD for GAN, NLP, LLM/VLM…

**Network augmentation**
- NetAug: augment the model to get extra supervision during training for tiny models

**Comment by surisurikim** [comment]
- KD for object detection
    - Detection에서는 출력 feature map의 크기나 형태가 다를 수 있으므로, 1x1 convolution을 사용하여 Teacher와 Student의 feature map을 같은 형태로 맞춘 뒤 차이를 비교함 -> 이때 Hint feature를 정렬한 후, L2 Loss를 사용하여 Teacher와 Student 간 feature를 유사하게 만듦.
    - Classification 출력에서는 Soft Label(Pₜ)을 통해 Weighted Cross Entropy Loss를 사용함.
    - 여기서 foreground(객체)와 background(배경) 클래스에 서로 다른 가중치를 적용
        - foreground class: 정확하게 예측해야 하므로 손실에 더 많은 비중
        - background class: 다소 틀려도 허용 가능하므로 손실에 덜 민감
    - 44p 수식 Regression (Bounding Box 위치 예측)에 대한 Distillation - Teacher의 성능을 상한선으로 삼아 Student가 일정 기준까지 따라오도록 유도
- NLP
    - LLM 경량화에서 pruning과 distillation을 결합하는 접근이 보편적인 관행이 되었다고 설명
    - **57페이지 도식 무슨 소린지 좀 모르겠음.. 보충 설명 필요**
- 68~69p 큰 모델은 overfitting, 작은 모델은 underfitting 문제가 있음 -> Network Augmentation을 통해 작은 모델의 underfitting 문제 해결하고자 함
- NetAug
    -  NetAug는 KD와 별개의 기법이며, 병행 시 더 좋은 결과를 낸다 (orthogonal)

**Comment by sw930718** [comment]
**Knowledge Distillation 소개** 
   - 기존문제사항 : 제한된 하드웨어 리소스, tiny 모델 성능 한계
   - KD 설명 : 학생을 선생과 일치시키려는 목적으로, teacher logits과 student logits 사이의 증류손실(distillation loss)도입. 
      물론 학생의 분류 loss값도 계산. 두 loss 값을 이용해서 student가 teacher의 결과와 유사해지게 만듦 
      
      예시) 고양이/개 : 분류 손실 [0,1] 값, 증류 손실 [0.4, 0.6]값 
      - 증류 손실 : T(temperature)를 이용해, 값을 조정해 어느정도 유사한지 뭐와 더 가까운지 등의 거리를 학습함
      
      KD =  $\frac{\exp(\frac{z_i}{T})}{\ \sum_j exp(\frac{z_i}{T}) } $
   
**무엇을 match 시킬것인지** 
- output logits   
   :  teacher와 student의 출력 logits 으로 loss 계산
- intermediate weights / features  
   : teacher와 student의 중간 가중치/중간 features 등 중간 텐서로 loss 계산
- gradients  
   : teacher와 student의 attention maps으로 loss 계산 

**증류할 대상이 뭔지**  
- Offline Distillation
   : 교사 모델을 먼저 훈련시킨 다음, 학생 모델을 가르침 
- Online Distillation
   : 교사 모델, 학생 모델을 함께 훈련시킴 
- Self-Distillation
   : 학생 모델이 자기 자신으로부터 스스로 배움

**Comment by tjdtnsu** [comment]
- logits, intermediate weights, features, gradients (attention map), sparsity patterns, relational information (그룹으로 묶어서 그룹 간 loss 비교, feature vector 만들어서 samples 단, model 단으로 나누기)로 kd 할 수 있음
- self distillation: step k로 갈수록 k-1 모델을 distill하는 것.
- online distillation: 두 student 모델 쓰기
- object detction kd: 여러 loss들 간으로 할 수 있음. regression to classification으로 바꾸어서 distribution으로도 loss 가능
- segmentation도 중간 feature 쓰기 가능
- GAN kd: generator에 student 모델 두기
- attention transfer: attention map을 kd
- network augmentation: 중간에 feature map을 추가해서 gradient를 그쪽으로만 흘려주는 것. dropblock은 큰 모델에만 효과가 있는 것인것에 반해 이것은 반대임.
